### PR TITLE
Introduce apply functionality for ParameterWrappers

### PR DESCRIPTION
**Pull request by jaosch**
ParameterWrappers are now a specified version of AbstractUnwrappables that must directly wrap around a jax. Array. They implement the `apply` method, which returns an updated instance of themselves. This can help prevent zero gradients 

TODO:
- [ ] Create tests
- [ ] Update documentation

### CODE REVIEW

File: klax/_wrappers.py [hunk 1]
```diff
@@ -1,47 +1,259 @@
...
+# ===-----------------------------------------------------------------------===#
+
+
+def contains_unwrappables(pytree):
+    """Check if a pytree contains unwrappables."""
+
+    def _is_unwrappable(leaf):
+        return isinstance(leaf, AbstractUnwrappable)
+
+    leaves = jtu.tree_leaves(pytree, is_leaf=_is_unwrappable)
```
Comment (Drenderer) [inline klax/_wrappers.py:258]: `jax.tree_util.tree_leaves` is deprecated, use `jax.tree.leaves` instead.

Comment (jaosch) [reply klax/_wrappers.py]: Done

File: klax/_wrappers.py [hunk 2]
```diff
@@ -1,47 +1,259 @@
...
+    def __init__(self, parameter: Array):
         # Ensure that the parameter fulfills the constraint initially
-        self.parameter = self._non_neg(px.unwrap(parameter))
+        self.parameter = self._non_neg(parameter)
 
     def _non_neg(self, x: Array) -> Array:
         return jnp.maximum(x, 0)
 
     def unwrap(self) -> Array:
         return self._non_neg(self.parameter)
```
Comment (Drenderer) [inline klax/_wrappers.py:240]: Having the unwrap not be the identity (i.e. return parameter directly) is not necessary here. It's not wrong, since it will not change the gradients, but it also contributes nothing really.
Was implemented to still have `unwrap` be the only call necessary to make a model callable? I think to be compatible with arbitrary ArrayWrappers we will need a function like `finalize` anyway, that combines combines `apply` and `unwrap` into one function call anyway.

File: klax/_wrappers.py [hunk 3]
```diff
@@ -1,47 +1,259 @@
...
+
+    def apply(self) -> "NonNegative":
+        return NonNegative(self._non_neg(self.parameter))
```
Comment (Drenderer) [inline klax/_wrappers.py:243]: Is creating a new Object the same as eqx.pytree_at? I'm interested to know if there are any performance differences between the two.

File: tests/conftest.py [hunk 4]
```diff
@@ -28,15 +27,18 @@ def getwrap():
...
-        def __init__(self, parameter: Array | px.AbstractUnwrappable[Array]):
-            self.parameter = jnp.zeros_like(px.unwrap(parameter))
+        def __init__(self, parameter: Array):
+            self.parameter = jnp.zeros_like(parameter)
 
         def unwrap(self) -> Array:
             return jnp.zeros_like(self.parameter)
 
+        def apply(self) -> "Wrapper":
+            return Wrapper(self.parameter)
```
Comment (Drenderer) [inline tests/conftest.py:40]: Why not just return self? Wouldn't that be cheaper than creating a new instance?

File: klax/_wrappers.py [hunk 5]
```diff
@@ -1,47 +1,259 @@
...
+        return (
+            NonTrainable(leaf)
+            if eqx.is_inexact_array(leaf) or isinstance(leaf, ArrayWrapper)
+            else leaf
+        )
+
+    return jax.tree.map(
+        f=_map_fn,
+        tree=tree,
+        is_leaf=lambda x: isinstance(x, NonTrainable),
```
[the remark this answers is not in the document]
Comment (Drenderer) [reply klax/_wrappers.py:140]: So `is_leaf=lambda x: isinstance(x, (NonTrainable, ArrayWrapper)),`

File: klax/_wrappers.py [hunk 6]
```diff
@@ -1,47 +1,259 @@
...
+
+    ``ArrayWrapper`` is a specialized version of ``paramax.AbstractUnwrappable``
+    that returns an updated version of itself upon applying. ``ArrayWrapper``
+    cannot be nested but is fully compatible with paramax's unwrapping functionality.
+    """
+
+    def __init__(self, parameter: Array):
         raise NotImplementedError("To be implemented by derived classes")
 
     @abstractmethod
```
[the remark this answers is not in the document]
Comment (Drenderer) [reply klax/_wrappers.py:178]: by default I mean
```
def apply(self) -> Self:
   return self
```

File: klax/_wrappers.py (hunk 3 above)
Comment (jaosch) [reply klax/_wrappers.py:243]: Good question. My understandig is that `eqx.Module` cannot be modified directly, therefore I would assume that `eqx.tree_at` creates a copy as well.

File: tests/conftest.py (hunk 4 above)
Comment (jaosch) [reply tests/conftest.py:40]: See previous answer. The parameter cannot simply be mutated in place.

File: klax/_wrappers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaosch) [reply klax/_wrappers.py:178]: I would be fine with creating a default.

File: klax/_wrappers.py [hunk 7]
```diff
@@ -1,47 +1,259 @@
...
+
+class ArrayWrapper(AbstractUnwrappable[Array]):
+    """An abstract class representing array wrappers.
+
+    ``ArrayWrapper`` is a specialized version of ``paramax.AbstractUnwrappable``
+    that returns an updated version of itself upon applying. ``ArrayWrapper``
+    cannot be nested but is fully compatible with paramax's unwrapping functionality.
+    """
+
+    def __init__(self, parameter: Array):
```
[the remark this answers is not in the document]
Comment (jaosch) [reply klax/_wrappers.py:175]: ok

File: klax/_wrappers.py [hunk 8]
```diff
@@ -1,47 +1,259 @@
...
+        >>> params = klax.NonNegative(-1 * jnp.ones(3))
+        >>> klax.apply(("abc", 1, params))
+        ('abc', 1, Array([0., 0., 0.], dtype=float32))
+    """
+
+    def _apply(tree, *, include_self: bool):
+        def _map_fn(leaf):
+            if isinstance(leaf, ArrayWrapper):
+                # Unwrap subnodes, then itself
+                return _apply(leaf, include_self=False).apply()
```
[the remark this answers is not in the document]
Comment (jaosch) [reply klax/_wrappers.py:209]: Yes it is not necessary. I just left it because it doesn't cost anything, but it is effectively not used.

File: klax/_wrappers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jaosch) [reply klax/_wrappers.py:140]: My understanding is that we want `non_negative` to create wrappers around the array within an `ArrayWrapper`. Therefore `ArrayWrapper` is added to line 134 but not line 141.

File: klax/_wrappers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaosch) [reply klax/_wrappers.py:178]: I would prefer to have `ArrayWrapper` be fully abstract. This makes the behavior of derived classes more explicit in the sense that I don't have to look at `ArrayWrapper` to understand what is happening upon applying.

File: klax/_wrappers.py (hunk 3 above)
Comment (jaosch) [reply klax/_wrappers.py:243]: I will change it to `eqx.tree_at` anyway.

File: tests/conftest.py (hunk 4 above)
Comment (jaosch) [reply tests/conftest.py]: I decided to use `eqx.tree_at` instead.

File: docs/source/klax.rst [hunk 9]
```diff
@@ -34,8 +34,17 @@ Wrappers and constraints
...
+    unwrap
+    Parameterize
+    NonTrainable
+    non_trainable
+    Symmetric
+    SkewSymmetric
+    Constraint
+    apply
     NonNegative
+    finalize
```
Comment (jaosch) [inline docs/source/klax.rst:47]: `contains_unwrappables` and `contains_constraints` is missing here.

File: klax/_wrappers.py [hunk 10]
```diff
@@ -12,92 +12,428 @@
...
     def __init__(self, parameter: Array):
-        if px.contains_unwrappables(parameter):
-            warn("Wrapping SkewSymmetric around wrapped parameters might result in unexpected behaviour.")
+        """
+        Args:
+            parameter: Wrapped matrix as array of shape (..., N, N).
+        """
+        if contains_unwrappables(parameter):
+            #FIXME: This warning should be an error or not exist at all.
+            warn(
```
Comment (jaosch) [inline klax/_wrappers.py:237]: I my opinion there should be no warning here. Either it is an error or the user is warned in the documentation. Otherwise, if a user makes the deliberate choice to wrap `SkewSymmetric` round `Unwrappable` there is a warning, even though it is intentional.

I am strongly opposed warnings in favor of code style.

File: klax/_wrappers.py [hunk 11]
```diff
@@ -12,92 +12,428 @@
...
     def __init__(self, parameter: Array):
-        if px.contains_unwrappables(parameter):
-            warn("Wrapping Symmetric around wrapped parameters might result in unexpected behaviour.")
+        """
+        Args:
+            parameter: To be wrapped matrix array of shape (..., N, N).
+        """
+        #FIXME: This warning should be an error or not exist at all.
+        if contains_unwrappables(parameter):
+            warn(
```
Comment (jaosch) [inline klax/_wrappers.py:276]: See the previous comment. Warnings are no good style in my opinion, because they send ambiguous messages.

File: klax/_wrappers.py [hunk 12]
```diff
@@ -12,92 +12,428 @@
...
+    """
+
+    parameter: Array
+
+    @staticmethod
+    def _non_neg(x: Array) -> Array:
+        return jnp.maximum(x, 0)
+
+    def unwrap(self) -> Array:
+        return self._non_neg(self.parameter)
```
Comment (jaosch) [inline klax/_wrappers.py:380]: This should just return the parameter without modification.

**Review by jaosch — COMMENTED** [review]
Looks good in general, but need some refinement here and there.
